rfcs/2020-09-29-4155-aws-s3-source.md

```diff
@@ -0,0 +1,187 @@
+SQS for bucket notifications.
+
+It will not cover the [proposed polling
+strategy](https://github.com/timberio/vector/issues/1017#issuecomment-694287111),
+but will leave space for it to be added later in a backwards compatible fashion.
+
+It will not cover parsing of events within the objects. Vector's current
+approach for this is to delegate to transforms. For example, it is likely we'll
+want to add a transform to extract CloudTrail events stored as objects in S3.
+
+## Motivation
+
+There are some types of logs that AWS can write to S3. For example:
+
+* CloudTrail logs (can also go to CloudWatch Logs but with a size limitation)
+* AWS load balancer access logs (only S3)
+* VPC flow logs (can also go to CloudWatch logs)
+
+Even if some AWS service logs _can_ go to CloudWatch Logs, I think it will be
+common that operators will want to just send them to S3 as it is cheaper storage
+if they don't need to regularly access them.
+
+## Internal Proposal
+
+A new `aws_s3` source will be added that will poll SQS for bucket notifications
+and ingest new objects from S3. It will wait to delete the SQS message until the
+object has been fully processed to guarantee "at least" once delivery to the
+pipeline (though it can still be lost after that until we tackle [end-to-end
+acks](https://github.com/timberio/vector/issues/3922).
+
+This approach will allow ingestion to be easily horizontally scaled by spinning
+up extra `vector` instances, relying on SQS to handle the fan-out.
+
+## Doc-level Proposal
+
+We will create a user guide based on
+[AWS's](https://docs.aws.amazon.com/AmazonS3/latest/user-guide/enable-event-notifications.html#s3-event-notification-destinations)
+for configuring bucket notifications to publish object created events to SQS.
+
+The new source configuration will look like:
+
+```toml
+[sources.my_source_id]
+  # General
+  type = "aws_s3" # required
+  bucket = "my-bucket" # required
+  region = "us-east-1" # required, required when endpoint = ""
+  compression = "auto" # optional; compression format of the objects; one of: ["auto", "none", "gzip" "lz4" "snappy" "zstd"]; default
+
+  strategy = "sqs" # optional, default, one of ["sqs"]
+
+  sqs.queue_url = "https://sqs.us-east-1.amazonaws.com/1234/test-s3-queue"
+  sqs.poll_secs = 30 # minimum poll interval, optional, default
+  sqs.visibility_timeout_secs = 300 # default visibility timeout for SQS message; if vector does not process the message in this time, the SQS message will be available to be reprocessed
+  sqs.delete_message = true # whether to delete the message after processing; false is useful for debugging; default
+```
+
+[A
+comment](https://github.com/timberio/vector/issues/1017#issuecomment-699125610)
+mentions desiring the ability to filter objects by prefix, but I think this is
+better configured when setting up bucket notifications to avoid publishing
+notifications for these objects at all.
+
```

Comment:
How about metrics?  One incredibly useful feature of this would be to see : 

- The rate of logs-files being processed. 
- The rate of new messages(logs) arriving in SQS 

So that I can quickly calculate if the Vector ingest is keeping up with the load, or if I need to introduce more replicas.